src/java.desktop/share/native/libfontmanager/HBShaper_Panama.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include <stdlib.h>
+#include "hb.h"
+#include "hb-jdk-p.h"
+#include "hb-ot.h"
+#include "scriptMapping.h"
+
+static float euclidianDistance(float a, float b)
+{
+    float root;
+    if (a < 0) {
+        a = -a;
+    }
+
+    if (b < 0) {
+        b = -b;
+    }
+
+    if (a == 0) {
+        return b;
+    }
+
+    if (b == 0) {
+        return a;
+    }
+
+    /* Do an initial approximation, in root */
+    root = a > b ? a + (b / 2) : b + (a / 2);
+
+    /* An unrolled Newton-Raphson iteration sequence */
+    root = (root + (a * (a / root)) + (b * (b / root)) + 1) / 2;
+    root = (root + (a * (a / root)) + (b * (b / root)) + 1) / 2;
+    root = (root + (a * (a / root)) + (b * (b / root)) + 1) / 2;
+
+    return root;
+}
+
+#define TYPO_KERN 0x00000001
+#define TYPO_LIGA 0x00000002
+#define TYPO_RTL  0x80000000
+
+JDKEXPORT int jdk_hb_shape(
+     float ptSize,
+     float *matrix,
+     void* pFace,
+     unsigned short *chars,
+     int len,
+     int script,
+     int offset,
+     int limit,
+     int baseIndex,
+     float startX,
+     float startY,
+     int flags,
+     int slot,
+     hb_font_get_nominal_glyph_func_t nominal_fn,
```

Comment:
It shouldn't be necessary to pass all the functions here. Note that the upcalls are now effectively static (due to the use of scope values). It would be more efficient to create the array of functions once and for all in Java code (either directly, by creating a memory segment and storing all the function pointers in there, or indirectly, by calling the `_hb_jdk_get_font_funcs` native function). But, we don't need to create a function array each time we call the `shape` function (as all the functions in the array are going to be the same after all). If you do that, you can replace all the `_fn` parameters in here with a single function pointer array parameter.